Calculate oriented bounding box from bounding region for 3D tiles of Cesium's OSM buildings

sraimund opened the pull request:
Here the ported code from Cesium to fix the calculation of the bounding box from a region on the WGS84 ellipsoid (see issue https://github.com/visgl/loaders.gl/issues/3144). The function signature fromRegion() in oriented-bounding-box.ts (math.gl/culling) is similar to the previous createObbFromRegion function in bounding-volume.ts (loaders.gl/tiles). The other functions are nearly a 1:1 mapping from Cesium, except that these are now class methods.

Test:
https://stackblitz.com/edit/stackblitz-starters-pode72re (takes ~2mins to build and load)

Notes:
- I could only build the culling module when adding `// @ts-ignore // eslint-disable-next-line import/no-unresolved` above `import {Ellipsoid} from '@math.gl/geospatial';` in oriented-bounding-box.ts and `import {Ellipsoid} from '@math.gl/geospatial';` in ellipsoid-tangent-plane.ts.

- I'm not sure whether the 'lint' command worked correctly since it only showed some warnings for the dev-docs.

Commit:
ported methods:
- mod
- zeroToTwoPi
- negativePiToPi

Commit:
partly ported class 'EllipsoidTangentPlane' from Cesium

Commit:
ported constant from Cesium:
- PI

Commit:
ported methods from Cesium:
- multiplyByVector
- multiplyByScale

Commit:
ported methods from Cesium:
- fromRegion
- fromPlaneExtents

Commit:
partly ported class 'Rectangle' from Cesium

Commit:
partly ported class 'Ray' from Cesium

Commit:
ported method from Cesium:
- intersectWithRay (originally in IntersectionTests)

Commit:
export ported classes

Commit:
added geospatial module for Rectangle calculations

Commit:
fix: reuse OBBs properties

ibgreen commented on modules/core/src/classes/matrix3.ts:
```diff
@@ -202,6 +203,56 @@ export class Matrix3 extends Matrix {
+   * @param {Vector3} result The object onto which to store the result.
+   * @returns {Vector3} The modified result parameter.
+   */
+  multiplyByVector(cartesian: Vector3, result?: Vector3): Vector3 {
```
This is the same as `m3.transform(v3)` below? The name comes from viewing the matrix as a transform on vectors.

sraimund replied:
It's indeed the same. I removed the method https://github.com/visgl/math.gl/pull/40/commits/2cb1e6b04b91a37fb10e16999f714a81b6f716d5

ibgreen commented on modules/core/src/classes/matrix3.ts:
```diff
@@ -202,6 +203,56 @@ export class Matrix3 extends Matrix {
+   * @param {Matrix3} result The object onto which to store the result.
+   * @returns {Matrix3} The modified result parameter.
+   */
+  multiplyByScale(scale: Vector3, result?: Matrix3): Matrix3 {
```
This method is inherited from `Matrix3 < Matrix < MathArray`. Called `m3.scale(v3)`

sraimund replied:
I now reused the parent scale method, see https://github.com/visgl/math.gl/pull/40/commits/801904feb5f95986fd582ad4b6c4f7f6a3b86caa The overridden scale method in Matrix3 multiplies it with a Vector2. There I'm not sure if adding another check whether the array length is 2 or 3 impacts the performance in case the method is called many times.

ibgreen commented on modules/core/src/lib/common.ts:
```diff
@@ -122,6 +123,55 @@ export function degrees(
+/**
+ * The modulo operation that also works for negative dividends.
+ *
+ * @param {number} m The dividend.
```
Nit: Avoid adding types in the TSDoc, they are inferred from the typescript types.

sraimund replied:
Done. See https://github.com/visgl/math.gl/pull/40/commits/2222cc563ec9e765ef80113d5e569e8d86dccb02

ibgreen commented on modules/culling/src/lib/bounding-volumes/oriented-bounding-box.ts:
```diff
@@ -98,6 +134,178 @@ export class OrientedBoundingBox implements BoundingVolume {
+   * @returns {OrientedBoundingBox} The modified result parameter or a new OrientedBoundingBox instance if none was provided.
+   */
+  // eslint-disable-next-line max-statements
+  fromRegion(region: number[]): OrientedBoundingBox {
```
Nit: From tree shaking (bundle size) perspective, it would be preferable to offer this as a global function rather than a method on the OBB class.  That way we could also avoid this file from growing too big.

sraimund replied:
Done. See https://github.com/visgl/math.gl/commit/5c2bd422ae2cd09a7a22a21254f85364a2361e6b. I renamed it to makeOrientedBoundingBoxFromRegion() similarly to the other function names in the algorithms folder.

ibgreen commented on modules/culling/src/lib/bounding-volumes/oriented-bounding-box.ts:
```diff
@@ -98,6 +134,178 @@ export class OrientedBoundingBox implements BoundingVolume {
+   * Computes an OrientedBoundingBox that bounds a region on the surface of the WGS84 ellipsoid.
+   * There are no guarantees about the orientation of the bounding box.
+   *
+   * @param {number[]} region The cartographic region ([west, south, east, north, minimum height, maximum height])
```
Nit: Remove type annotations here.

sraimund replied:
Done. See https://github.com/visgl/math.gl/commit/2222cc563ec9e765ef80113d5e569e8d86dccb02

ibgreen commented on modules/core/src/index.ts:
```diff
@@ -58,6 +58,9 @@ export {
   // math.gl "GLSL"-style functions
   radians,
   degrees,
+  mod,
```
At least a trivial, minimal test case for every global export is normally expected.

sraimund replied:
Done. See https://github.com/visgl/math.gl/pull/40/commits/0f89190af5a3f063514092aca032a77b1a1abc65

sraimund replied:
In bounding-box-from-region.spec.ts, however, I added only two tests since there are too many to port from Cesium. Moreover, these tests are based on a unit sphere and math.gl supports only a WGS84 ellipsoid currently.

ibgreen commented on modules/culling/src/lib/ellipsoid-tangent-plane.ts:
```diff
@@ -0,0 +1,102 @@
+const scratchDirection = new Vector3();
+
+/** A plane tangent to the WGS84 ellipsoid at the provided origin */
+export class EllipsoidTangentPlane {
```
Are there any tests for this class that could be ported? (even if most are commented out, still nice to have them)

sraimund replied:
Done. See https://github.com/visgl/math.gl/commit/0f89190af5a3f063514092aca032a77b1a1abc65

ibgreen commented on modules/culling/src/lib/plane.ts:
```diff
@@ -86,4 +87,34 @@ export class Plane {
+   * @param {Vector3} [result] The object onto which to store the result.
+   * @returns {Vector3} The intersection point or undefined if there is no intersections.
+   */
+  intersectWithRay(ray: Ray, result?: Vector3): Vector3 {
```
Nit: Might have preferred as a global function to avoid pulling in `Ray` for every app.

sraimund replied:
Done. See https://github.com/visgl/math.gl/pull/40/commits/5c2bd422ae2cd09a7a22a21254f85364a2361e6b

ibgreen commented on modules/geospatial/src/rectangle.ts:
```diff
@@ -0,0 +1,68 @@
+import {Vector3, negativePiToPi, _MathUtils} from '@math.gl/core'
+
+/** A two dimensional region specified as longitude and latitude coordinates. */
+export class Rectangle {
```
Nit: Maybe this should be called `LngLatRectangle` given its special handling of geospatial coordinates (it is certainly not a generic Rectangle class).

sraimund replied:
I renamed it to GeoRectangle https://github.com/visgl/math.gl/pull/40/commits/ae5cef2092d52c51db98ff73b99822f54ad69cdc (similarly to https://doc.qt.io/qt-6/qgeorectangle.html). Concerning your proposition, I would be a bit confused if it is called LngLatRectangle or LatLngRectangle.

ibgreen reviewed (approved):
@sraimund Thanks for doing this relatively extensive work, that was definitely a good amount of code!

Apart from some smaller nits (optional suggestion) in my comments, the main missing pieces are docs and tests, at least for the publicly exported functions.

I understand if you have limits on the amount of time you can put on this PR, just let me know what you can do, and then we will get it landed and published.

ibgreen committed:
Merge branch 'master' into master

ibgreen committed:
chore: Bump lerna to 4.2.0-alpha.0. Fix CI. Modernize dependencies. (#41)

ibgreen committed:
feat(geospatial): Add makeOBBFromRegion() (#43)

Co-authored-by: Raimund Schnürer <raimund.schnurer@epfl.ch>

ibgreen committed:
v4.2.0-alpha.1

dependabot[bot] committed:
build(deps): bump nanoid from 3.3.7 to 3.3.8 (#39)

Signed-off-by: dependabot[bot] <support@github.com>

ibgreen committed:
chore: Fix CI workflows (#46)

ibgreen committed:
v4.2.0-alpha.2

ibgreen committed:
chore: Fix CI release action (#47)

ibgreen committed:
v4.2.0-alpha.3

Commit:
remove types from TSdoc

Commit:
reuse parent method

Commit:
global functions makeOrientedBoundingBoxFromRegion() and intersectPlaneWithRay() to prevent large files and improve tree shaking

Commit:
rename Rectangle to GeoRectangle

Commit:
tests ported

Commit:
minimal tests added

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
replace multiplyByVector() by transform()

sraimund commented:
I've also updated https://stackblitz.com/edit/stackblitz-starters-pode72re with a new build from the recent commits. 

Do the docs need to be manually added or can they be somehow automatically transferred from TSdoc?

ibgreen commented:
@sraimund This PR needs to be rebased. Looks like we may have duplicated some of the required changes.

Commit:
moved from culling to geospatial

Commit:
use normalizeAngle() instead of negativePiToPi()

Commit:
use normalizeAngle() instead of negativePiToPi() and zeroToTwoPi(), safeMod() instead of mod()

Commit:
moved from culling to geospatial

Commit:
remove duplicate methods

Commit:
adjust exports and imports

Commit:
removed in favor of geo-rectangle

Commit:
removed in favor of keeping method in culling module

Commit:
tests for EllipsoidTangentPlane

Commit:
amend tsdoc

Commit:
adjust version

sraimund commented:
> @sraimund This PR needs to be rebased. Looks like we may have duplicated some of the required changes.

I rebased it on my fork, removed duplicate methods and resolved the conflicts:

Accepted your changes:
- export normalizeAngle() instead of negativePiToPi() and zeroToTwoPi()
- move EllipsoidTangentPlane to geospatial module

Kept my changes:
- export makeOrientedBoundingBoxFromRegion() from culling module (see discussion above)
- class name GeoRectangle instead of LntLatRectangle (see discussion above)
